provider/pinecone.go

```diff
@@ -0,0 +1,446 @@
+	base := api.getVectorOperationURL(indexName, "vectors/upsert")
+	payload := &upsertRequest{
+		Vectors: []vectorElement{
+			{
+				ID:     api.generateDeterministicID(id),
+				Values: vector,
+				Metadata: metadataElement{
+					"id": id,
+				},
+			},
+		},
+		Namespace: namespace,
+	}
+	body, err := api.request(http.MethodPost, base, payload, http.StatusOK)
+	if err != nil {
+		return err
+	}
+	var response upsertResponse
+	err = json.Unmarshal(body, &response)
+	if err != nil {
+		return err
+	}
+	if response.UpsertedCount != 1 {
+		return fmt.Errorf("expected 1 upserted count, got %d", response.UpsertedCount)
+	}
+	return nil
+}
+
+// https://docs.pinecone.io/reference/fetch
+func (api pineconeAPI) fetch(indexName, namespace, id string) ([]float32, error) {
+	base, err := url.Parse(api.getVectorOperationURL(indexName, "vectors/fetch"))
+	if err != nil {
+		return nil, err
+	}
+	vectorID := api.generateDeterministicID(id)
+	params := url.Values{}
+	params.Add("ids", vectorID)
+	params.Add("namespace", namespace)
+	base.RawQuery = params.Encode()
+	body, err := api.request(http.MethodGet, base.String(), nil, http.StatusOK)
+	if err != nil {
+		return nil, err
+	}
+	var response fetchResponse
+	err = json.Unmarshal(body, &response)
+	if err != nil {
+		return nil, err
+	}
+	vector, ok := response.Vectors[vectorID]
+	if !ok {
+		return nil, fmt.Errorf("vector not found")
+	}
+	return vector.Values, nil
+}
+
+// https://docs.pinecone.io/reference/query
+func (api pineconeAPI) query(indexName, namespace string, vector []float32, k int64) ([]string, error) {
+	base := api.getVectorOperationURL(indexName, "query")
+	payload := &queryRequest{
+		Vector:    vector,
+		Namespace: namespace,
+		TopK:      k,
+		// Given the original/raw id for the vector is stored in the vector's metadata map,
+		// it's necessary to return the metadata for each result so that the original id can be
+		// returned to the user as the UUID5 representation has no meaning outside of Pinecone.
+		IncludeMetadata: true,
+	}
+	body, err := api.request(http.MethodPost, base, payload, http.StatusOK)
+	if err != nil {
+		return nil, err
+	}
+	var response queryResponse
+	err = json.Unmarshal(body, &response)
+	if err != nil {
+		return nil, err
+	}
+	results := make([]string, k)
+	for i, result := range response.Matches {
+		results[i] = result.Metadata["id"]
+	}
+	return results, nil
+}
+
+func (api pineconeAPI) getIndexOperationURL(operation string) string {
+	subdomain := fmt.Sprintf("controller.%s", api.config.Environment)
+	return fmt.Sprintf(api.baseURLTemplate, subdomain, operation)
+}
+
+func (api pineconeAPI) getVectorOperationURL(indexName, operation string) string {
+	subdomain := fmt.Sprintf("%s-%s.svc.%s", indexName, api.config.ProjectID, api.config.Environment)
+	return fmt.Sprintf(api.baseURLTemplate, subdomain, operation)
+}
+
+func (api pineconeAPI) request(method, url string, payload interface{}, expectedStatus int) ([]byte, error) {
+	var reader io.Reader
```

Comment:
Does pinecone not have a Go client?

Reply:
It appears they began work on [`go-pinecone`](https://github.com/pinecone-io/go-pinecone) but discontinue work (see the note in the README). Technically, the API docs offer client code generation, but it wasn't really that useful and I figured I'd implement the wrapper while learning the API.